activerecord/lib/active_record/inheritance.rb

```diff
@@ -198,12 +198,12 @@ def type_condition(table = arel_table)
       # If this is a StrongParameters hash, and access to inheritance_column is not permitted,
       # this will ignore the inheritance column and return nil
       def subclass_from_attributes?(attrs)
-        attribute_names.include?(inheritance_column) && (attrs.is_a?(Hash) || attrs.respond_to?(:permitted?))
+        attribute_names.include?(inheritance_column) && (attrs.is_a?(Hash) || attrs.respond_to?(:permitted?) || columns_hash[inheritance_column].default)
       end
 
       def subclass_from_attributes(attrs)
         attrs = attrs.to_h if attrs.respond_to?(:permitted?)
-        subclass_name = attrs.with_indifferent_access[inheritance_column]
+        subclass_name = (attrs || {}).with_indifferent_access[inheritance_column] || columns_hash[inheritance_column].default
```

Comment:
ah, I see. It's because we modify `subclass_from_attributes?` in a weird way. Why not keep `subclass_from_attributes?` as is but modify this part:

[`def new`](https://github.com/kuldeepaggarwal/rails/blob/fix-STI-default-type/activerecord/lib/active_record/inheritance.rb#L54-L56)
```ruby
        if subclass_from_attributes?(attrs)
          subclass = subclass_from_attributes(attrs)
        else
          subclass = columns_hash[inheritance_column].default
        end
```

This way the methods still do what their name suggests and it should be obvious at a glance how the subclass is determined.